Conveyor belt using behavior tree

boraini opened the pull request:
Change WalkBehavior so it checks if the NPC is on top of an item with an use type of move: true, and behave differently in that case.
- [ ] The entities might be going too fast.

boraini committed:
add find block utility

boraini committed:
add move item use type

boraini committed:
add conveyor belt behavior

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/WalkBehavior.java:
```diff
@@ -19,7 +23,32 @@ public WalkBehavior(@JacksonInject Npc entity) {
+        Zone zone = entity.getZone();
+
+        // regular walk behavior if the zone is not known
+        if (zone != null) {
```
Check is redundant: `zone` will never be `null` in this location.

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/zone/Zone.java:
```diff
@@ -574,7 +574,37 @@ && isBlockSolid(x1, y1, false)) {
         return false;
     }
-    
+
+    /**
+     * Find block with item occupying the block position that satisfies the predicate.
+     * Closer blocks are prioritized in row major order.
+     */
+    public Block findBlock(int x, int y, Layer layer, Predicate<Item> predicate) {
+        Block block;
+        Item item;
+
+        for (int i = 0; i >= -3; i--) {
+            for (int j = 0; j <= 2; j++) {
+                int x1 = x + i;
+                int y1 = y + j;
+
+                if (!areCoordinatesInBounds(x1, y1) || !isChunkLoaded(x1, y1)) {
+                    continue;
+                }
+
+                block = getBlock(x1, y1);
+                item = block.getItem(Layer.FRONT);
+
+                if (item.getBlockWidth() > Math.abs(i) && item.getBlockHeight() > Math.abs(j)
+                        && predicate.test(item)) {
+                    return block;
+                }
+            }
+        }
+
+        return null;
+    }
+
```
I'd rather have this function moved to the `WalkBehavior` or `Npc` class and have it serve as a function to find the block the entity is standing on.

boraini replied:
I agree with all the other comments but what is wrong with having the findBlock function here? I think it is relevant on the zone domain, and could be used to refactor a lot more logic in the future since testing for large blocks occupying a block position is important. Also, could I do something with metablocks instead (I don't think so at all)?

kuroppoi replied:
It's fine to keep it there, but then I would suggest updating `isBlockSolid(x, y, checkAdjacents)` to use this function in order to keep duplicate code to a minimum.

Edit: To answer your second question, conveyor belts aren't metablocks, so no.
It's not really appropriate to query metablocks in a scenario such as this, anyway.
Metablocks typically exist for one of two reasons:
- To store extra data (loot containers, player-configurable signs, etc.)
- To quickly find the location(s) of important blocks without loading chunks (zone teleporters, protectors, etc.)

Storing metablocks can be rather expensive because of this, so I would advise against turning items into metablocks without a good reason.

boraini replied:
So my suspicion about metablocks being similar to tile entities from Minecraft was correct. I will keep this in mind. Also, I will refactor isBlockSolid. I have thought about it but wouldn't change it before asking.

boraini replied:
Hi. Can you please check https://github.com/boraini/brainwine/pull/7 before I add it here?

kuroppoi replied:
Yeah looks good, feel free to include it here. I'll edit the nitpicky things myself later.

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/WalkBehavior.java:
```diff
@@ -19,7 +23,32 @@ public WalkBehavior(@JacksonInject Npc entity) {
+            }
+        }
+
+        if (movingSurfacePower == null) {
+            setAnimation("walk");
+            entity.move(entity.getDirection().getId(), 0, animation);
+        } else {
+            setAnimation("idle");
+            entity.setDirection(movingSurfacePower > 0 ? FacingDirection.EAST : FacingDirection.WEST);
+            entity.move(movingSurfacePower.intValue(), 0, animation);
+        }
```
The `setAnimation` calls here will change the _behavior's_ animation, which should not be done.
Use `entity.setAnimation(name)`,`entity.move(x, y, animation)` or `entity.move(x, y, speed, animation)` to update the NPC's animation.

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/WalkBehavior.java:
```diff
@@ -19,7 +23,32 @@ public WalkBehavior(@JacksonInject Npc entity) {
+        } else {
+            setAnimation("idle");
+            entity.setDirection(movingSurfacePower > 0 ? FacingDirection.EAST : FacingDirection.WEST);
+            entity.move(movingSurfacePower.intValue(), 0, animation);
```
`movingSurfacePower` should probably be used as the NPC's speed rather than its movement offset.
Speed can be updated using `entity.setSpeed(speed)`, but using `entity.move(x, y, speed, animation)` to set everything in one go is preferred.

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/WalkBehavior.java:
```diff
@@ -19,7 +23,32 @@ public WalkBehavior(@JacksonInject Npc entity) {
     public boolean behave() {
-        entity.move(entity.getDirection().getId(), 0, animation);
+        // null if the npc should walk. Power value with sign if on moving surface.
+        Float movingSurfacePower = null;
```
Non-primitive numbers are scary. I suggest simply checking if `block` is `null` instead.

boraini committed:
refactor code for the conveyor belt

boraini committed:
use proper animations for conveyor belt

boraini committed:
refactor isBlockSolid to use the new findBlock method

boraini committed:
Merge pull request #7 from boraini/is-block-solid-refactoring

isBlockSolid refactoring

boraini commented:
Another thing... Do we want the NPCs to look towards west and east randomly while going on the conveyors? It seems to be the case in the Ruby server.

kuroppoi commented:
There seems to be a bug regarding this, actually.
Apparently, conveyor belts give entities the special ability to walk through solid walls.

kuroppoi committed:
Simplify `isSolid()` function

kuroppoi committed:
Use behavior animation for regular walking behavior

kuroppoi committed:
Small changes

kuroppoi committed:
Shorten code

kuroppoi committed:
Remove unneeded `null` check

The server defaults items to `Item.AIR` (or it's supposed to, at the very least.)

kuroppoi commented:
I took the liberty of changing a few nitpicky things, but the bug remains.
It seems to happen because the conveyor belt can move entities in directions that they aren't facing, which bypasses the solid wall check.
An easy solution would be to implement an additional solid wall check in the conveyor belt movement section.

boraini commented:
> I took the liberty of changing a few nitpicky things, but the bug remains. It seems to happen because the conveyor belt can move entities in directions that they aren't facing, which bypasses the solid wall check. An easy solution would be to implement an additional solid wall check in the conveyor belt movement section.

I was afraid of introducing bugs so I didn't do much code shortening myself. I will look into clipping into the walls today since I am done with university stuff.

boraini committed:
Handle walls in front of conveyed entities

boraini committed:
Split conveyor belt behavior into a new partial behavior

boraini commented:
Hello. I have added a fix for the wall clipping problem. I also changed the architecture of the conveyor belt behavior so it can be composed into other behavior types easily. You can see the work for that in https://github.com/boraini/brainwine/pull/8. Tell me if you want it.

kuroppoi commented:
Probably a better idea yeah, feel free to include it.

boraini committed:
Merge pull request #8 from boraini/split-conveyor-belt-behavior

Split conveyor belt behavior into a new partial behavior

boraini commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/Behavior.java:
```diff
@@ -46,6 +47,7 @@
     @Type(name = "walk", value = WalkBehavior.class),
     @Type(name = "fall", value = FallBehavior.class),
     @Type(name = "turn", value = TurnBehavior.class),
+    @Type(name = "conveyor_belt", value = ConveyorBeltBehavior.class),
```
Make sure to check if this is fine please.

kuroppoi committed:
Change conveyor behavior

kuroppoi committed:
Conveyor behavior should get priority over idle behavior

kuroppoi committed:
Add conveyor behavior to crawler & digger behavior trees

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/Entity.java:
```diff
@@ -338,7 +338,11 @@ public int getSizeX() {
         return sizeY;
     }
-        
+    
+    public void setDirection(int direction) {
+        setDirection(direction > 0 ? FacingDirection.EAST : direction < 0 ? FacingDirection.WEST : this.direction);
+    }
+    
```
Utility function

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/Npc.java:
```diff
@@ -254,11 +254,18 @@ public void move(int x, int y, String animation) {
     }
     
     public void move(int x, int y, float speed, String animation) {
+        move(x, y, speed, animation, true);
+    }
+    
+    public void move(int x, int y, float speed, String animation, boolean changeDirection) {
         this.speed = speed;
-        direction = x > 0 ? FacingDirection.EAST : x < 0 ? FacingDirection.WEST : direction;
         moveX = x;
         moveY = y;
         
+        if(changeDirection) {
+            setDirection(x);
+        }
+        
```
More utility

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/composed/WalkerBehavior.java:
```diff
@@ -27,6 +28,8 @@ public WalkerBehavior(Npc entity) {
 
     @Override
     protected void addChildren(Map<String, Object> config) {
+        addChild(ConveyorBehavior.class, config);
+        
```
Conveyor behavior should always go first because idle behavior can cancel out the movement

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/ConveyorBehavior.java:
```diff
@@ -0,0 +1,50 @@
+        return conveyorBlock != null;
+    }
+    
+    public void setAnimation(String animation) {
+        this.animation = animation;
+    }
```
I removed the `@JsonAlias` annotation since it's only there for out-of-the-box compatibility with the game configuration (conveyor behavior isn't standard so it is not necessary)

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/ConveyorBehavior.java:
```diff
@@ -0,0 +1,50 @@
+    
+    @Override
+    public boolean canBehave() {
+        // Check if entity is standing on a conveyor belt
+        conveyorBlock = entity.getZone().findBlock(entity.getBlockX(), entity.getBlockY() + 1, block -> block.getFrontItem().hasUse(ItemUseType.MOVE));
+        return conveyorBlock != null;
```
Behavior is instanced so we can grab the conveyor block here and save it instead of doing it twice

kuroppoi commented on gameserver/src/main/java/brainwine/gameserver/entity/npc/behavior/parts/ConveyorBehavior.java:
```diff
@@ -0,0 +1,50 @@
+        int direction = conveyorBlock.getFrontMod() == 0 ? 1 : -1;
+        float movingSurfacePower = conveyorBlock.getFrontItem().getPower();
+        
+        // Randomly change direction to match the conveyor belt's
+        if(Math.random() < 0.333) {
+            entity.setDirection(direction);
+        }
```
Not really fond of this one myself but it does the job well enough I'd say
